flash/data/data_source.py

```diff
@@ -0,0 +1,263 @@
+
+    def load_data(self,
+                  data: DATA_TYPE,
+                  dataset: Optional[Any] = None) -> Union[Sequence[Mapping[str, Any]], Iterable[Mapping[str, Any]]]:
+        """Loads entire data from Dataset. The input ``data`` can be anything, but you need to return a Mapping.
+
+        Example::
+
+            # data: "."
+            # output: [("./cat/1.png", 1), ..., ("./dog/10.png", 0)]
+
+            output: Mapping = load_data(data)
+
+        """
+        return data
+
+    def load_sample(self, sample: Mapping[str, Any], dataset: Optional[Any] = None) -> Any:
+        """Loads single sample from dataset"""
+        return sample
+
+    def to_datasets(
+        self,
+        train_data: Optional[DATA_TYPE] = None,
+        val_data: Optional[DATA_TYPE] = None,
+        test_data: Optional[DATA_TYPE] = None,
+        predict_data: Optional[DATA_TYPE] = None,
+    ) -> Tuple[Optional[BaseAutoDataset], ...]:
+        train_dataset = self.generate_dataset(train_data, RunningStage.TRAINING)
+        val_dataset = self.generate_dataset(val_data, RunningStage.VALIDATING)
+        test_dataset = self.generate_dataset(test_data, RunningStage.TESTING)
+        predict_dataset = self.generate_dataset(predict_data, RunningStage.PREDICTING)
+        return train_dataset, val_dataset, test_dataset, predict_dataset
+
+    def generate_dataset(
+        self,
+        data: Optional[DATA_TYPE],
+        running_stage: RunningStage,
+    ) -> Optional[Union[AutoDataset, IterableAutoDataset]]:
+        is_none = data is None
+
+        if isinstance(data, Sequence):
+            is_none = data[0] is None
+
+        if not is_none:
+            from flash.data.data_pipeline import DataPipeline
+
+            mock_dataset = MockDataset()
+            with CurrentRunningStageFuncContext(running_stage, "load_data", self):
+                load_data: Callable = getattr(
+                    self, DataPipeline._resolve_function_hierarchy(
+                        "load_data",
+                        self,
+                        running_stage,
+                        DataSource,
+                    )
+                )
+                parameters = signature(load_data).parameters
+                if len(parameters) > 1 and "dataset" in parameters:  # TODO: This was DATASET_KEY before
+                    data = load_data(data, mock_dataset)
+                else:
+                    data = load_data(data)
+
+            if has_len(data):
+                dataset = AutoDataset(data, self, running_stage)
+            else:
+                dataset = IterableAutoDataset(data, self, running_stage)
+            dataset.__dict__.update(mock_dataset.metadata)
+            return dataset
+
+
+class DefaultDataSources(LightningEnum):
+
+    PATHS = "paths"
+    NUMPY = "numpy"
+    TENSOR = "tensor"
+    CSV = "csv"
+    JSON = "json"
+
+    # TODO: Create a FlashEnum class???
+    def __hash__(self) -> int:
+        return hash(self.value)
```

Comment:
Not sure, the hash in lightning enum uses the name for some reason. If that is not deliberate then we can upstream this to lightning, otherwise we could make a flashenum class that does what we need.

Reply:
@tchaton thoughts?